src/Utilities/FlowAnalysis/Analysis/PropertySetAnalysis/PropertySetAnalysisContext.cs

```diff
@@ -0,0 +1,141 @@
+            this.PropertyToSetFlag = propertyToSetFlag;
+            this.IsNullPropertyFlagged = isNullPropertyFlagged;
+            this.MethodNamesToCheckForFlaggedUsage = methodNamesToCheckForFlaggedUsage;
+        }
+
+        public static PropertySetAnalysisContext Create(
+            AbstractValueDomain<PropertySetAbstractValue> valueDomain,
+            WellKnownTypeProvider wellKnownTypeProvider,
+            ControlFlowGraph controlFlowGraph,
+            ISymbol owningSymbol,
+            InterproceduralAnalysisKind interproceduralAnalysisKind,
+            bool pessimisticAnalysis,
+            PointsToAnalysisResult pointsToAnalysisResultOpt,
+            Func<PropertySetAnalysisContext, PropertySetAnalysisResult> getOrComputeAnalysisResult,
+            string typeToTrackMetadataName,
+            bool isNewInstanceFlagged,
+            string propertyToSetFlag,
+            bool isNullPropertyFlagged,
+            ImmutableHashSet<string> methodNamesToCheckForFlaggedUsage)
+
+        {
+            return new PropertySetAnalysisContext(
+                valueDomain, wellKnownTypeProvider, controlFlowGraph, owningSymbol, interproceduralAnalysisKind,
+                pessimisticAnalysis, pointsToAnalysisResultOpt, getOrComputeAnalysisResult, null,
+                null,
+                typeToTrackMetadataName,
+                isNewInstanceFlagged,
+                propertyToSetFlag,
+                isNullPropertyFlagged,
+                methodNamesToCheckForFlaggedUsage);
+        }
+
+        public override PropertySetAnalysisContext ForkForInterproceduralAnalysis(
+            IMethodSymbol invokedMethod,
+            ControlFlowGraph invokedCfg,
+            IOperation operation,
+            PointsToAnalysisResult pointsToAnalysisResultOpt,
+            CopyAnalysisResult copyAnalysisResultOpt,
+            InterproceduralBinaryFormatterAnalysisData interproceduralAnalysisData)
+        {
+            Debug.Assert(pointsToAnalysisResultOpt != null);
+            Debug.Assert(copyAnalysisResultOpt == null);
+
+            // Do not invoke any interprocedural analysis more than one level down.
+            // We only care about analyzing validation methods.
+            return new PropertySetAnalysisContext(
+                ValueDomain, WellKnownTypeProvider, invokedCfg, invokedMethod, InterproceduralAnalysisKind.None,
+                PessimisticAnalysis, pointsToAnalysisResultOpt, GetOrComputeAnalysisResult, ControlFlowGraph,
+                interproceduralAnalysisData,
+                this.TypeToTrackMetadataName,
+                this.IsNewInstanceFlagged,
+                this.PropertyToSetFlag,
+                this.IsNullPropertyFlagged,
+                this.MethodNamesToCheckForFlaggedUsage);
+        }
+
+        /// <summary>
+        /// Metadata name of the type to track.
+        /// </summary>
+        public string TypeToTrackMetadataName { get; }
```

Comment:
>TypeToTrackMetadataName [](start = 22, length = 23)

It seems that currently only one security analyzer is using this flow analysis, so the current implementation is fine. However, if you see more use cases with different TypeToTrackMetadataName, it might be better to switch this to a dictionary from TypeToTrackMetadataName to an object wrapping the below properties, and same for the resulting hazardous usages map. This way you can avoid multiple different analysis passes to compute the result. 

Do you think CA3075 will also use a similar property set analysis to avoid false positives such as [https://github.com/dotnet/roslyn-analyzers/issues/1591](https://github.com/dotnet/roslyn-analyzers/issues/1591)?

Reply:
Actually, I take that back. I looked at your current analyzer and it invokes flow analysis only if the specific banned methods are invoked. So, it is unlikely that multiple analyzers can share analysis pass/result, as the cases where they need flow analysis will vary vastly. Your current implementation is fine :). However, if you do add more property set analyses in future, like for CA3075, it would be good to share code, especially parts of operation visitor.

---
In reply to: [223056981](https://github.com/dotnet/roslyn-analyzers/pull/1838#discussion_r223056981) [](ancestors = 223056981)